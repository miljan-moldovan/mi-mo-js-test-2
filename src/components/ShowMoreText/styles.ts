import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  panelTopRemarks: {
    textAlign: 'justify',
    color: '#110A24',
    fontSize: 11,
    padding: 5,
    backgroundColor: '#F1F1F1',
    fontFamily: 'Roboto',
  },
  panelInfoShowMoreText: {
    color: '#115ECD',
    fontSize: 10,
    fontFamily: 'Roboto',
    backgroundColor: 'transparent',
    textAlign: 'center',
  },
  panelInfoShowMore: {
    borderColor: '#CACBCF',
    borderWidth: 1,
    borderRadius: 10,
    backgroundColor: '#FFFFFF',
    width: 100,
    position: 'absolute',
    bottom: -12,
  },
  buttonShowView: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  panelTopLineLeft: {
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    paddingHorizontal: 2,
    backgroundColor: 'transparent',
    flex: 1,
  },
  panelTopLine: {
    flexDirection: 'row',
    justifyContent: 'center',
    width: '90%',
    alignItems: 'center',
    padding: 5,
    backgroundColor: '#F1F1F1',
  },
  showMoreTextView: {
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export default styles;
