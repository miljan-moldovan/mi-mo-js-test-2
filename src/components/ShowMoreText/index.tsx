import * as React from 'react';
import { Icons } from 'react-native-fontawesome';
import SalonTouchableOpacity from '../SalonTouchableOpacity';
import styles from './styles';
import { Text, View } from 'react-native';

type IProps = {
  text: string,
  maxLength?: number,
};

const ShowButton = (props) => {
  return (
    <View style={styles.buttonShowView}>
      <SalonTouchableOpacity
        style={styles.panelInfoShowMore}
        onPress={props.onPressShowMore}
      >
        <Text style={styles.panelInfoShowMoreText}>
          {props.title}
        </Text>
      </SalonTouchableOpacity>
    </View>
  );
};


class ShowMoreText extends React.Component<IProps> {
  static defaultProps = {
    maxLength: 140,
  };

  state = {
    showMore: true,
    text: (this.props.text.length > this.props.maxLength)
      ? this.props.text.substring(0, this.props.maxLength)
      : this.props.text,
  };

  componentWillReceiveProps(newProps) {
    if (this.props.text !== newProps.text || this.props.maxLength !== newProps.maxLength) {
      this.setState(
        {
          text: (newProps.text.length > this.props.maxLength)
            ? newProps.text.substring(0, this.props.maxLength)
            : newProps.text,
        },
      );
    }
  }

  onPressShowMore = () => {
    return this.setState({
      showMore: false,
      text: this.props.text,
    });
  };
  onPressShowLess = () => {
    return this.setState({
      showMore: true,
      text: this.props.text.substring(0, this.props.maxLength),
    });
  };

  render() {
    return (
      (

        <View style={styles.panelTopLine}>
          <View style={styles.panelTopLineLeft}>
            <View style={{ justifyContent: 'flex-start', width: '100%' }}>
              <Text
                style={styles.panelTopRemarks}
              >
                {this.state.text}
              </Text>
            </View>
            {this.props.text.length > this.props.maxLength &&
            (
              this.state.showMore ?
                <ShowButton onPressShowMore={this.onPressShowMore} title="SHOW MORE" /> :
                <ShowButton onPressShowMore={this.onPressShowLess} title="SHOW LESS" />
            )
            }
          </View>
        </View>

      )
    );
  }
}

export default ShowMoreText;
