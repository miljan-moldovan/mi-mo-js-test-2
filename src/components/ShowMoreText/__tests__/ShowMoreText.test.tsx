import * as React from 'react';
import { mount, configure } from 'enzyme';
import ShowMoreText from '../index';

describe('<ShowMoreText />', () => {
  let defaultProps;

  // just for remove error with depend using react-dom for render. Need it before use adapter for RN
  const origConsole = console.error;
  const mountComponent = (props = defaultProps) => {
    return mount(<ShowMoreText {...props} />);
  };

  beforeEach(() => {
    // just for remove error with depend using react-dom for render. Need it before use adapter for RN
    console.error = () => {};
    defaultProps = {
      maxLength: 20,
      text:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut laoreet quam dolor. \
        Praesent nunc ante, interdum eget elementum rhoncus, vulputate et ante. \
        Aenean a semper urna, at maximus ligula posuere.',
    };
  });
  afterEach(() => {
    // just for remove error with depend using react-dom for render. Need it before use adapter for RN
    console.error = origConsole;
  });

  it('renders correctly', () => {
    expect(mountComponent()).toMatchSnapshot();
  });

  describe('Should properly cut down text if its length exceeds max length', () => {
    it('text exceeds max length', () => {
      const renderedText = mountComponent()
        .find('Text')
        .first()
        .text();
      expect(renderedText.length).toBe(defaultProps.maxLength);
    });
    it('text does not (!) exceed max length', () => {
      defaultProps.text = 'hello this is dog';
      const renderedText = mountComponent()
        .find('Text')
        .first()
        .text();
      expect(renderedText.length).toBe(defaultProps.text.length);
    });
  });

  describe('Should update state when text from props change', () => {
    it('new text is the same as old', () => {
      const component = mountComponent();
      const oldState = component.state();
      component.setProps({ ...defaultProps });
      expect(component.state()).toEqual(oldState);
    });
    it('new text is not the same', () => {
      const component = mountComponent();
      const oldState = component.state();
      component.setProps({ ...defaultProps, text: 'hello this is dog' });
      expect(component.state()).not.toEqual(oldState);
    });
    it('new text is cut down to maxLength', () => {
      const component = mountComponent();
      component.setProps({ ...defaultProps, text: 'abcdefghijklmnopqrstuvwxyz' });
      expect(component.state('text').length).toEqual(defaultProps.maxLength);
    });
  });

  describe('Should properly render `show more/less` button', () => {
    it('text exceeds max length => SHOULD render button', () => {
      const button = mountComponent().exists('SalonTouchableOpacity(TouchableOpacity)');
      expect(button).toEqual(true);
    });
    it('text does not (!) exceed max length => SHOULD NOT render button', () => {
      defaultProps.text = 'hello this is dog';
      const button = mountComponent().exists('SalonTouchableOpacity(TouchableOpacity)');
      expect(button).toEqual(false);
    });
  });

  describe('Buttons should properly toggle state', () => {
    it('state toggles', () => {
      const component = mountComponent();

      component.instance().onPressShowMore();
      expect(component.state('showMore')).toEqual(false);

      component.instance().onPressShowLess();
      expect(component.state('showMore')).toEqual(true);
    });
  });
});
