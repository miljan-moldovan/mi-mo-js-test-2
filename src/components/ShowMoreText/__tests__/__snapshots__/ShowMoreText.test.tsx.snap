// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShowMoreText /> renders correctly 1`] = `
<ShowMoreText
  maxLength={20}
  text="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut laoreet quam dolor.         Praesent nunc ante, interdum eget elementum rhoncus, vulputate et ante.         Aenean a semper urna, at maximus ligula posuere."
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F1F1F1",
        "flexDirection": "row",
        "justifyContent": "center",
        "padding": 5,
        "width": "90%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#F1F1F1",
          "flexDirection": "row",
          "justifyContent": "center",
          "padding": 5,
          "width": "90%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
            "paddingHorizontal": 2,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "center",
              "paddingHorizontal": 2,
            }
          }
        >
          <View
            style={
              Object {
                "justifyContent": "flex-start",
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "justifyContent": "flex-start",
                  "width": "100%",
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "backgroundColor": "#F1F1F1",
                    "color": "#110A24",
                    "fontFamily": "Roboto",
                    "fontSize": 11,
                    "padding": 5,
                    "textAlign": "justify",
                  }
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  style={
                    Object {
                      "backgroundColor": "#F1F1F1",
                      "color": "#110A24",
                      "fontFamily": "Roboto",
                      "fontSize": 11,
                      "padding": 5,
                      "textAlign": "justify",
                    }
                  }
                >
                  Lorem ipsum dolor si
                </Text>
              </Text>
            </View>
          </View>
          <ShowButton
            onPressShowMore={[Function]}
            title="SHOW MORE"
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                  }
                }
              >
                <SalonTouchableOpacity(TouchableOpacity)
                  onPress={[Function]}
                  style={
                    Object {
                      "backgroundColor": "#FFFFFF",
                      "borderColor": "#CACBCF",
                      "borderRadius": 10,
                      "borderWidth": 1,
                      "bottom": -12,
                      "position": "absolute",
                      "width": 100,
                    }
                  }
                >
                  <TouchableOpacity
                    activeOpacity={0.2}
                    onPress={[Function]}
                    style={
                      Object {
                        "backgroundColor": "#FFFFFF",
                        "borderColor": "#CACBCF",
                        "borderRadius": 10,
                        "borderWidth": 1,
                        "bottom": -12,
                        "position": "absolute",
                        "width": 100,
                      }
                    }
                  >
                    <AnimatedComponent
                      accessible={true}
                      isTVSelectable={true}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Array [
                          Object {
                            "backgroundColor": "#FFFFFF",
                            "borderColor": "#CACBCF",
                            "borderRadius": 10,
                            "borderWidth": 1,
                            "bottom": -12,
                            "position": "absolute",
                            "width": 100,
                          },
                          Object {
                            "opacity": 1,
                          },
                        ]
                      }
                    >
                      <View
                        accessible={true}
                        isTVSelectable={true}
                        onResponderGrant={[Function]}
                        onResponderMove={[Function]}
                        onResponderRelease={[Function]}
                        onResponderTerminate={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        style={
                          Object {
                            "backgroundColor": "#FFFFFF",
                            "borderColor": "#CACBCF",
                            "borderRadius": 10,
                            "borderWidth": 1,
                            "bottom": -12,
                            "opacity": 1,
                            "position": "absolute",
                            "width": 100,
                          }
                        }
                      >
                        <View
                          accessible={true}
                          isTVSelectable={true}
                          onResponderGrant={[Function]}
                          onResponderMove={[Function]}
                          onResponderRelease={[Function]}
                          onResponderTerminate={[Function]}
                          onResponderTerminationRequest={[Function]}
                          onStartShouldSetResponder={[Function]}
                          style={
                            Object {
                              "backgroundColor": "#FFFFFF",
                              "borderColor": "#CACBCF",
                              "borderRadius": 10,
                              "borderWidth": 1,
                              "bottom": -12,
                              "opacity": 1,
                              "position": "absolute",
                              "width": 100,
                            }
                          }
                        >
                          <Text
                            accessible={true}
                            allowFontScaling={true}
                            ellipsizeMode="tail"
                            style={
                              Object {
                                "backgroundColor": "transparent",
                                "color": "#115ECD",
                                "fontFamily": "Roboto",
                                "fontSize": 10,
                                "textAlign": "center",
                              }
                            }
                          >
                            <Text
                              accessible={true}
                              allowFontScaling={true}
                              ellipsizeMode="tail"
                              style={
                                Object {
                                  "backgroundColor": "transparent",
                                  "color": "#115ECD",
                                  "fontFamily": "Roboto",
                                  "fontSize": 10,
                                  "textAlign": "center",
                                }
                              }
                            >
                              SHOW MORE
                            </Text>
                          </Text>
                        </View>
                      </View>
                    </AnimatedComponent>
                  </TouchableOpacity>
                </SalonTouchableOpacity(TouchableOpacity)>
              </View>
            </View>
          </ShowButton>
        </View>
      </View>
    </View>
  </View>
</ShowMoreText>
`;
